# Solving Systems of Linear Equations with NumPy - Lab

## Introduction 

Now we have gathered all the required skills needed to solve systems of linear equations. We saw why need to calculate inverses of matrices, followed by matrix multiplication to figure the values of unknown variables. 

Following exercises present some problems that can be converted into a system of linear equations. Solve the problems for the values of unknown variables. 

## Objectives
You will be able to:

* Build a system of equations from given problem by creating matrix and vectors in NumPy
* Calculate the inverse and dot-product to get the values for unknown variables.
* Solve linear equations using `numpy.linalg.solve()`

## Exercise 1

A coffee shop is having a sale on coffee and tea. 

On day 1, 29 bags of coffee and 41 bags of tea was sold, giving a total of 490 dollars.

On day 2, they sold 23 bags of coffee and 41 bags of tea, for which customers paid a total of 448 dollars.  

How much does each bag cost?

In [16]:
# Solution:
import numpy as np
X = np.array([[29,41],[23,41]])
X.shape

(2, 2)

In [20]:
Y = np.array([[490],[448]])
Y.shape

(2, 1)

In [21]:
# Explain your answer in terms of prices
A = np.linalg.solve(X,Y)
A

array([[7.],
       [7.]])

In [22]:
A.shape

(2, 1)

Each bag costs 7$ 

## Exercise 2:

The cost of admission to a popular music concert was 162 dollars for 12 children and 3 adults. 

The admission was 122 dollars for 8 children and 3 adults in another music concert. 

How much was the admission for each child and adult?

In [8]:
# Create a system of equations
import numpy as np
A = np.array([[12,3],[8,3]])  # number of children and adults
print(A.shape)

Y = np.array([162,122])
print(Y.shape)

X = np.linalg.solve(A,Y)
print(X)

(2, 2)
(2,)
[10. 14.]


In [9]:
# Describe your result
# price per child is 10 and price per adult is 14

## Exercise 3:

You want to make a certain kind of tropical punch, using bananas, oranges, and papayas. 

Suppose you don't know how many of each to put in the punch, but you know that there are seven pieces of fruit in the mix, and there are twice as many oranges as bananas. 

You also know that the seven pieces of fruit cost \$5.25, 

where bananas cost \$.50 each, oranges cost \$.75 each, and papayas cost \$1.25 each.

How many bananas, oranges and papayas would you need to make this punch ?

In [25]:
# Create a system of equations

A = np.array([[(0.5+0.75*2),1.25],[3,1]])
A.shape

(2, 2)

In [29]:
Y = np.array([[5.25],[7]])
Y.shape

(2, 1)

In [31]:
# Solving
X = np.linalg.solve(A,Y)
X

array([[2.],
       [1.]])

## Answer Explanation

Solved for:


EQ1 :    0.5X1 + 0.75X2 + 1.25X3 = 5.25 where X2 = 2*X1 (because twice as much oranges than bananas)


And


EQ2 : X1 + X2 + X3 = 7   => 3X1 + X3 = 7


X1 = # of bananas is 2


X2 = # of oranges is 2*X1 = 4


X3 = # of papayas is 1


## Exercise 4

A landlord owns 3 houses, a 1-bedroom, a 2-bedroom , and a 3-bedroom house. 

The total rent he receives is 1240 dollars. 

He needs to make some repairs, where it costs 10% of the 1-bedroom house’s rent for its repairs, 20% of the 2-bedroom for its repairs, and 30% of the 3-bedroom house's rent for its repairs.  The total repair bill for all three houses was $276. 

The 3-bedroom house's rent is twice the 1-bedroom house’s rent. 

How much is the individual rent for three houses?

In [38]:
# Create a system of equations
# Let Ri be the rent for apartment Ri where i is the # of bedrooms
# 0.1* R1 + 0.2 * R2 + 0.3 * R3 = 276
# R3 = 2* R1  meaning that 2R1 + 0R2 - R3 = 0
# R1 + R2 + R3 = 1240



In [39]:
# Explain your answer
A = np.array([[0.1,0.2,0.3],[2,0,-1],[1,1,1]])
A.shape


(3, 3)

In [42]:
Y = np.array([[276],[0],[1240]])
Y.shape

(3, 1)

In [43]:
np.linalg.solve(A,Y)

array([[280.],
       [400.],
       [560.]])

## Answer
Rent for apart1 = 280

Rent for apart2 = 400

Rent for apart3 = 560

## Summary
In this lab, we saw how to use Numpy to solve linear equations by taking inverses and matrix multiplication and also using numpy's `.solve()`. We shall now take these skills forward and see how we can define a simple regression problem using linear algebra and solve it with numpy, only using the skills that we have learnt so far. 